File: viewing-schedule/internal/tmdb/tmdb.go
package tmdb

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

const (
	baseURL = "https://api.themoviedb.org/3"
	imgURL  = "https://image.tmdb.org/t/p"
)

// Client is a small TMDB API client.
type Client struct {
	token string
	http  *http.Client
}

// New creates a new TMDB client. token may be empty (lookups will fail gracefully).
func New(token string) *Client {
	return &Client{
		token: token,
		http:  &http.Client{Timeout: 15 * time.Second},
	}
}

// Movie is the result of a TMDB lookup.
type Movie struct {
	ID          int64
	Title       string
	Year        string
	Overview    string
	Director    string
	Runtime     int
	Genres      []string
	Rating      float64
	Poster      string
	PosterLg    string
	Backdrop    string
	ReleaseDate string
	Tagline     string
}

// HasToken reports whether the client is configured with a token.
func (c *Client) HasToken() bool { return c.token != "" }

// ParseTitle extracts "name" and "year" from a string like "The Apartment (1960)".
func ParseTitle(s string) (name, year string) {
	re := regexp.MustCompile(`^(.+?)\s*\((\d{4})\)\s*$`)
	m := re.FindStringSubmatch(strings.TrimSpace(s))
	if len(m) != 3 {
		return strings.TrimSpace(s), ""
	}
	return strings.TrimSpace(m[1]), m[2]
}

// Lookup searches TMDB for a movie and returns enriched details.
func (c *Client) Lookup(ctx context.Context, rawTitle string) (*Movie, error) {
	if !c.HasToken() {
		return nil, errors.New("tmdb: no token configured")
	}
	name, year := ParseTitle(rawTitle)
	if name == "" {
		return nil, errors.New("tmdb: empty title")
	}

	params := url.Values{}
	params.Set("query", name)
	if year != "" {
		params.Set("year", year)
	}
	params.Set("include_adult", "false")

	var search struct {
		Results []struct {
			ID           int64   `json:"id"`
			Title        string  `json:"title"`
			Overview     string  `json:"overview"`
			PosterPath   string  `json:"poster_path"`
			BackdropPath string  `json:"backdrop_path"`
			VoteAverage  float64 `json:"vote_average"`
			ReleaseDate  string  `json:"release_date"`
		} `json:"results"`
	}
	if err := c.get(ctx, "/search/movie?"+params.Encode(), &search); err != nil {
		return nil, err
	}
	if len(search.Results) == 0 {
		return nil, fmt.Errorf("tmdb: no result for %q", rawTitle)
	}
	r := search.Results[0]

	var details struct {
		Runtime int `json:"runtime"`
		Genres  []struct {
			Name string `json:"name"`
		} `json:"genres"`
		Tagline string `json:"tagline"`
		Credits struct {
			Crew []struct {
				Name string `json:"name"`
				Job  string `json:"job"`
			} `json:"crew"`
		} `json:"credits"`
	}
	if err := c.get(ctx, fmt.Sprintf("/movie/%d?append_to_response=credits", r.ID), &details); err != nil {
		return nil, err
	}

	director := ""
	for _, c := range details.Credits.Crew {
		if c.Job == "Director" {
			director = c.Name
			break
		}
	}

	genres := make([]string, 0, len(details.Genres))
	for _, g := range details.Genres {
		genres = append(genres, g.Name)
	}

	m := &Movie{
		ID:          r.ID,
		Title:       r.Title,
		Year:        year,
		Overview:    r.Overview,
		Director:    director,
		Runtime:     details.Runtime,
		Genres:      genres,
		Rating:      r.VoteAverage,
		ReleaseDate: r.ReleaseDate,
		Tagline:     details.Tagline,
	}
	if r.PosterPath != "" {
		m.Poster = imgURL + "/w342" + r.PosterPath
		m.PosterLg = imgURL + "/w500" + r.PosterPath
	}
	if r.BackdropPath != "" {
		m.Backdrop = imgURL + "/w780" + r.BackdropPath
	}
	if year == "" && len(r.ReleaseDate) >= 4 {
		m.Year = r.ReleaseDate[:4]
	}
	return m, nil
}

func (c *Client) get(ctx context.Context, path string, out any) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, baseURL+path, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+c.token)
	req.Header.Set("Accept", "application/json")
	res, err := c.http.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode/100 != 2 {
		return fmt.Errorf("tmdb: status %d for %s", res.StatusCode, path)
	}
	return json.NewDecoder(res.Body).Decode(out)
}

File: marquee/app/tmdb.py
"""TMDB client — search, trending, and film metadata with posters.

Falls back to bundled fixtures when no API key is configured, so the UI is
always explorable.
"""
import httpx
from . import config, fixtures

BASE = "https://api.themoviedb.org/3"


def _client():
    headers = {"accept": "application/json"}
    params = {}
    if config.TMDB_BEARER:
        headers["Authorization"] = f"Bearer {config.TMDB_BEARER}"
    elif config.TMDB_API_KEY:
        params["api_key"] = config.TMDB_API_KEY
    return httpx.Client(base_url=BASE, headers=headers, params=params, timeout=12)


def img(path, size="w500"):
    if not path:
        return None
    if path.startswith("http"):
        return path
    return f"{config.TMDB_IMG}/{size}{path}"


def normalize(m: dict) -> dict:
    """Normalize a TMDB movie payload to Marquee's film shape."""
    year = None
    rd = m.get("release_date") or ""
    if len(rd) >= 4 and rd[:4].isdigit():
        year = int(rd[:4])
    director = None
    credits = m.get("credits") or {}
    for c in credits.get("crew", []):
        if c.get("job") == "Director":
            director = c.get("name")
            break
    genres = [g["name"] for g in m.get("genres", [])] if m.get("genres") else m.get("genre_names", [])
    return {
        "tmdb_id": m["id"],
        "title": m.get("title") or m.get("name"),
        "year": year,
        "poster": img(m.get("poster_path")),
        "backdrop": img(m.get("backdrop_path"), "w1280"),
        "overview": m.get("overview"),
        "runtime": m.get("runtime"),
        "director": director,
        "genres": genres,
        "tagline": m.get("tagline"),
        "vote": round(m.get("vote_average", 0), 1) if m.get("vote_average") else None,
    }


def search(query: str):
    if not config.TMDB_ENABLED:
        return fixtures.search(query)
    with _client() as c:
        r = c.get("/search/movie", params={"query": query, "include_adult": "false"})
        r.raise_for_status()
        return [normalize(m) for m in r.json().get("results", [])[:24]]


def trending():
    if not config.TMDB_ENABLED:
        return fixtures.trending()
    with _client() as c:
        r = c.get("/trending/movie/week")
        r.raise_for_status()
        return [normalize(m) for m in r.json().get("results", [])[:18]]


def movie(tmdb_id: int):
    if not config.TMDB_ENABLED:
        return fixtures.movie(tmdb_id)
    with _client() as c:
        r = c.get(f"/movie/{tmdb_id}", params={"append_to_response": "credits"})
        r.raise_for_status()
        return normalize(r.json())
